File: androidApp/src/main/kotlin/org/coreypett/fullstack/features/marketdetails/components/OrderBookSection.kt
package org.coreypett.fullstack.features.marketdetails.components

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.lerp
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.coreypett.fullstack.market.MR
import org.coreypett.fullstack.orderbook.model.LevelChange
import org.coreypett.fullstack.orderbook.model.OrderBookSide
import org.coreypett.fullstack.orderbook.presentation.OrderBookRowDisplay
import org.coreypett.fullstack.orderbook.presentation.OrderBookRowPair
import org.coreypett.fullstack.orderbook.presentation.OrderBookStatus
import org.coreypett.fullstack.orderbook.presentation.OrderBookViewState
import org.coreypett.fullstack.support.toComposeColor

@Composable
fun OrderBookSection(
    viewState: OrderBookViewState,
) {
    if (!viewState.hasRows) {
        if (viewState.status == OrderBookStatus.Failed) {
            EmptyState(message = viewState.centerMessage ?: "Loading order book")
        } else {
            OrderBookSkeleton()
        }
        return
    }

    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(0.dp),
    ) {
        OrderBookHeader()
        SpreadRow(
            spreadText = viewState.spreadText ?: "--",
            spreadPercentText = viewState.spreadPercentText ?: "--",
        )
        viewState.pairedRows.take(MaxOrderBookRows).forEach { pair ->
            PairedOrderBookRow(
                pair = pair,
            )
        }
    }
}

@Composable
private fun OrderBookSkeleton() {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(0.dp),
    ) {
        OrderBookHeader()
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(38.dp)
                .background(MR.colors.app_panel.toComposeColor())
                .padding(horizontal = 4.dp),
            contentAlignment = Alignment.Center,
        ) {
            SkeletonBlock(
                width = 116.dp,
                height = 16.dp,
                cornerRadius = 4.dp,
            )
        }
        repeat(8) { seed ->
            SkeletonOrderBookRow(seed = seed)
        }
    }
}

@Composable
private fun SkeletonOrderBookRow(
    seed: Int,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(30.dp)
            .background(MR.colors.app_background.toComposeColor()),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(OrderBookTextCenterGap),
    ) {
        Row(
            modifier = Modifier
                .weight(1f)
                .padding(start = OrderBookOuterPadding, end = OrderBookCenterPadding),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            SkeletonCell(
                width = skeletonSizeWidth(seed),
                alignment = Alignment.CenterStart,
            )
            SkeletonCell(
                width = skeletonPriceWidth(seed),
                alignment = Alignment.CenterEnd,
            )
        }
        Row(
            modifier = Modifier
                .weight(1f)
                .padding(start = OrderBookCenterPadding, end = OrderBookOuterPadding),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            SkeletonCell(
                width = skeletonPriceWidth(seed + 2),
                alignment = Alignment.CenterStart,
            )
            SkeletonCell(
                width = skeletonSizeWidth(seed + 1),
                alignment = Alignment.CenterEnd,
            )
        }
    }
}

@Composable
private fun RowScope.SkeletonCell(
    width: Dp,
    alignment: Alignment,
) {
    Box(
        modifier = Modifier.weight(1f),
        contentAlignment = alignment,
    ) {
        SkeletonBlock(
            width = width,
            height = 13.dp,
            cornerRadius = 3.dp,
        )
    }
}

@Composable
private fun OrderBookHeader() {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(OrderBookTextCenterGap),
    ) {
        Row(
            modifier = Modifier
                .weight(1f)
                .padding(start = OrderBookOuterPadding, end = OrderBookCenterPadding),
        ) {
            HeaderCell("Size", TextAlign.Start)
            HeaderCell("Price (Bid)", TextAlign.End)
        }
        Row(
            modifier = Modifier
                .weight(1f)
                .padding(start = OrderBookCenterPadding, end = OrderBookOuterPadding),
        ) {
            HeaderCell("Price (Ask)", TextAlign.Start)
            HeaderCell("Size", TextAlign.End)
        }
    }
}

@Composable
private fun RowScope.HeaderCell(
    text: String,
    textAlign: TextAlign,
) {
    Text(
        text = text,
        modifier = Modifier.weight(1f),
        color = MR.colors.text_tertiary.toComposeColor(),
        fontSize = 11.sp,
        textAlign = textAlign,
    )
}

@Composable
private fun PairedOrderBookRow(
    pair: OrderBookRowPair,
) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(30.dp),
    ) {
        DepthColumns(
            bid = pair.bid,
            ask = pair.ask,
            modifier = Modifier.matchParentSize(),
        )

        Row(
            modifier = Modifier.matchParentSize(),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(OrderBookTextCenterGap),
        ) {
            BidColumns(
                level = pair.bid,
                modifier = Modifier.weight(1f),
            )
            AskColumns(
                level = pair.ask,
                modifier = Modifier.weight(1f),
            )
        }
    }
}

@Composable
private fun DepthColumns(
    bid: OrderBookRowDisplay?,
    ask: OrderBookRowDisplay?,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(OrderBookDepthCenterGap),
    ) {
        Row(
            modifier = Modifier
                .weight(1f)
                .padding(start = OrderBookOuterPadding, end = OrderBookCenterPadding),
        ) {
            PriceDepth(
                level = bid,
                side = OrderBookSide.Bid,
            )
        }
        Row(
            modifier = Modifier
                .weight(1f)
                .padding(start = OrderBookCenterPadding, end = OrderBookOuterPadding),
        ) {
            PriceDepth(
                level = ask,
                side = OrderBookSide.Ask,
            )
        }
    }
}

@Composable
private fun RowScope.PriceDepth(
    level: OrderBookRowDisplay?,
    side: OrderBookSide,
) {
    val sideColor = if (side == OrderBookSide.Bid) MR.colors.bid.toComposeColor() else MR.colors.ask.toComposeColor()
    val targetDepth = level?.depthFraction?.coerceIn(0f, 1f) ?: 0f
    BoxWithConstraints(
        modifier = Modifier
            .weight(1f)
            .fillMaxHeight()
    ) {
        if (targetDepth > 0f) {
            Box(
                modifier = Modifier
                    .align(if (side == OrderBookSide.Bid) Alignment.CenterEnd else Alignment.CenterStart)
                    .width(maxWidth * targetDepth)
                    .fillMaxHeight()
                    .background(sideColor.copy(alpha = 0.11f)),
            )
        }
    }
}

@Composable
private fun BidColumns(
    level: OrderBookRowDisplay?,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier.padding(start = OrderBookOuterPadding, end = OrderBookCenterPadding),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        SizeText(
            level = level,
            side = OrderBookSide.Bid,
            textAlign = TextAlign.Start,
        )
        LevelText(
            text = level?.priceText,
            color = MR.colors.bid.toComposeColor(),
            textAlign = TextAlign.End,
        )
    }
}

@Composable
private fun AskColumns(
    level: OrderBookRowDisplay?,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier.padding(start = OrderBookCenterPadding, end = OrderBookOuterPadding),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        LevelText(
            text = level?.priceText,
            color = MR.colors.ask.toComposeColor(),
            textAlign = TextAlign.Start,
        )
        SizeText(
            level = level,
            side = OrderBookSide.Ask,
            textAlign = TextAlign.End,
        )
    }
}

@Composable
private fun RowScope.SizeText(
    level: OrderBookRowDisplay?,
    side: OrderBookSide,
    textAlign: TextAlign,
) {
    val sideColor = if (side == OrderBookSide.Bid) MR.colors.bid.toComposeColor() else MR.colors.ask.toComposeColor()
    val textColor = MR.colors.text_primary.toComposeColor()
    val flashProgress = remember(level?.rowKey) { Animatable(0f) }

    LaunchedEffect(level?.rowKey, level?.sizeText, level?.change, level?.sizeChangeFraction) {
        val startProgress = when {
            level == null -> 0f
            level.change == LevelChange.None -> 0f
            level.sizeChangeFraction < OrderBookFlashMinChangeFraction -> 0f
            level.change == LevelChange.Up -> OrderBookFlashUpTextProgress
            else -> OrderBookFlashDownTextProgress
        }
        if (startProgress > 0f) {
            flashProgress.snapTo(startProgress)
            flashProgress.animateTo(
                targetValue = 0f,
                animationSpec = tween(durationMillis = 140, easing = FastOutSlowInEasing),
            )
        } else {
            flashProgress.snapTo(0f)
        }
    }

    Box(
        modifier = Modifier
            .weight(1f)
            .fillMaxHeight(),
        contentAlignment = if (textAlign == TextAlign.End) Alignment.CenterEnd else Alignment.CenterStart,
    ) {
        Text(
            text = level?.sizeText ?: "",
            color = lerp(textColor, sideColor, flashProgress.value),
            fontSize = 13.sp,
            fontFamily = FontFamily.Monospace,
            textAlign = textAlign,
            maxLines = 1,
        )
    }
}

@Composable
private fun RowScope.LevelText(
    text: String?,
    color: Color,
    textAlign: TextAlign,
) {
    Text(
        text = text ?: "",
        modifier = Modifier.weight(1f),
        color = color,
        fontSize = 13.sp,
        fontFamily = FontFamily.Monospace,
        textAlign = textAlign,
        maxLines = 1,
    )
}

@Composable
private fun SpreadRow(
    spreadText: String,
    spreadPercentText: String,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(38.dp)
            .background(MR.colors.app_panel.toComposeColor())
            .padding(horizontal = 4.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(OrderBookSpreadItemGap, Alignment.CenterHorizontally),
    ) {
        Text(
            text = "Spread",
            color = MR.colors.text_tertiary.toComposeColor(),
            fontSize = 12.sp,
            maxLines = 1,
        )
        Text(
            text = spreadText,
            color = MR.colors.brand_orange.toComposeColor(),
            fontSize = 15.sp,
            fontWeight = FontWeight.SemiBold,
            fontFamily = FontFamily.Monospace,
            maxLines = 1,
        )
        Text(
            text = spreadPercentText,
            color = MR.colors.text_tertiary.toComposeColor(),
            fontSize = 12.sp,
            fontWeight = FontWeight.Medium,
            fontFamily = FontFamily.Monospace,
            maxLines = 1,
        )
    }
}

private val OrderBookTextCenterGap = 8.dp
private val OrderBookDepthCenterGap = 0.dp
private val OrderBookCenterPadding = 0.dp
private val OrderBookOuterPadding = 4.dp
private val OrderBookSpreadItemGap = 12.dp
private const val OrderBookFlashMinChangeFraction = 0.08f
private const val OrderBookFlashUpTextProgress = 0.88f
private const val OrderBookFlashDownTextProgress = 0.78f
private const val MaxOrderBookRows = 12

private val SkeletonPriceWidths = listOf(72.dp, 84.dp, 64.dp, 78.dp)
private val SkeletonSizeWidths = listOf(34.dp, 46.dp, 38.dp, 52.dp)

private fun skeletonPriceWidth(seed: Int): Dp = SkeletonPriceWidths[seed % SkeletonPriceWidths.size]

private fun skeletonSizeWidth(seed: Int): Dp = SkeletonSizeWidths[seed % SkeletonSizeWidths.size]

File: iosApp/iosApp/Features/MarketDetails/Components/OrderBookSection.swift
@preconcurrency import SharedLogic
import SwiftUI

struct OrderBookState {
    let status: OrderBookStatus
    let statusLabel: String
    let centerMessage: String?
    let spreadText: String?
    let spreadPercentText: String?
    let asks: [OrderBookRow]
    let bids: [OrderBookRow]
    let isLoading: Bool

    var hasRows: Bool {
        !asks.isEmpty || !bids.isEmpty
    }

    var pairedRows: [OrderBookLevelPair] {
        let visibleAsks = Array(asks.reversed())
        let rowCount = max(bids.count, visibleAsks.count)
        return (0..<rowCount).map { index in
            OrderBookLevelPair(
                bid: bids.indices.contains(index) ? bids[index] : nil,
                ask: visibleAsks.indices.contains(index) ? visibleAsks[index] : nil
            )
        }
    }

    static let preview = OrderBookState(
        status: .live,
        statusLabel: "Live",
        centerMessage: nil,
        spreadText: "$0.50",
        spreadPercentText: "0.001%",
        asks: [
            OrderBookRow(side: .ask, priceText: "$69,128", sizeText: "0.75", orderCountText: "1", depthFraction: 0.38, change: .none),
            OrderBookRow(side: .ask, priceText: "$69,127.50", sizeText: "1.00", orderCountText: "2", depthFraction: 0.5, change: .down),
            OrderBookRow(side: .ask, priceText: "$69,126", sizeText: "2.00", orderCountText: "4", depthFraction: 1.0, change: .none),
        ],
        bids: [
            OrderBookRow(side: .bid, priceText: "$69,125.50", sizeText: "1.25", orderCountText: "3", depthFraction: 1.0, change: .none),
            OrderBookRow(side: .bid, priceText: "$69,124", sizeText: "0.50", orderCountText: "1", depthFraction: 0.4, change: .up),
            OrderBookRow(side: .bid, priceText: "$69,123.50", sizeText: "0.25", orderCountText: "2", depthFraction: 0.2, change: .none),
        ],
        isLoading: false
    )

    static let loading = OrderBookState(
        status: .connecting,
        statusLabel: "Connecting",
        centerMessage: nil,
        spreadText: nil,
        spreadPercentText: nil,
        asks: [],
        bids: [],
        isLoading: true
    )

    init(shared: OrderBookViewState) {
        self.status = shared.status
        self.statusLabel = shared.statusLabel
        self.centerMessage = shared.centerMessage
        self.spreadText = shared.spreadText
        self.spreadPercentText = shared.spreadPercentText
        self.asks = shared.asks.map(OrderBookRow.init)
        self.bids = shared.bids.map(OrderBookRow.init)
        self.isLoading = shared.status == .connecting && !shared.hasRows
    }

    init(
        status: OrderBookStatus,
        statusLabel: String,
        centerMessage: String?,
        spreadText: String?,
        spreadPercentText: String?,
        asks: [OrderBookRow],
        bids: [OrderBookRow],
        isLoading: Bool
    ) {
        self.status = status
        self.statusLabel = statusLabel
        self.centerMessage = centerMessage
        self.spreadText = spreadText
        self.spreadPercentText = spreadPercentText
        self.asks = asks
        self.bids = bids
        self.isLoading = isLoading
    }
}

struct OrderBookRow: Hashable {
    let rowKey: String
    let side: OrderBookSide
    let priceText: String
    let sizeText: String
    let orderCountText: String
    let depthFraction: Double
    let sizeChangeFraction: Double
    let change: LevelChange

    init(shared: OrderBookRowDisplay) {
        self.rowKey = shared.rowKey
        self.side = shared.side
        self.priceText = shared.priceText
        self.sizeText = shared.sizeText
        self.orderCountText = shared.orderCountText
        self.depthFraction = Double(shared.depthFraction)
        self.sizeChangeFraction = Double(shared.sizeChangeFraction)
        self.change = shared.change
    }

    init(
        rowKey: String? = nil,
        side: OrderBookSide,
        priceText: String,
        sizeText: String,
        orderCountText: String,
        depthFraction: Double,
        sizeChangeFraction: Double = 0,
        change: LevelChange
    ) {
        self.rowKey = rowKey ?? "\(side)-\(priceText)"
        self.side = side
        self.priceText = priceText
        self.sizeText = sizeText
        self.orderCountText = orderCountText
        self.depthFraction = depthFraction
        self.sizeChangeFraction = sizeChangeFraction
        self.change = change
    }
}

struct OrderBookLevelPair {
    let bid: OrderBookRow?
    let ask: OrderBookRow?
}

struct OrderBookSection: View {
    let orderBook: OrderBookState

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            if orderBook.isLoading {
                OrderBookSkeleton()
                    .transition(.movingParts.blur.combined(with: .opacity))
            } else if orderBook.hasRows {
                OrderBookList(orderBook: orderBook)
                    .transition(.movingParts.blur.combined(with: .opacity))
            } else {
                CenterMessage(text: orderBook.centerMessage ?? "Connecting")
                    .frame(height: 220)
                    .transition(.movingParts.blur.combined(with: .opacity))
            }
        }
        .animation(.easeInOut(duration: 0.22), value: orderBook.isLoading)
    }
}

private struct OrderBookSkeleton: View {
    var body: some View {
        VStack(spacing: 0) {
            HeaderRow()
            SkeletonBlock(width: 116, height: 16, cornerRadius: 4)
                .frame(maxWidth: .infinity, minHeight: 38, alignment: .center)

            ForEach(0..<8, id: \.self) { index in
                SkeletonOrderBookRow(seed: index)
            }
        }
    }
}

private struct SkeletonOrderBookRow: View {
    let seed: Int

    var body: some View {
        HStack(spacing: orderBookTextCenterGap) {
            HStack(spacing: 0) {
                SkeletonBlock(width: sizeWidth(for: seed), height: 13, cornerRadius: 3)
                    .frame(maxWidth: .infinity, alignment: .leading)
                SkeletonBlock(width: priceWidth(for: seed), height: 13, cornerRadius: 3)
                    .frame(maxWidth: .infinity, alignment: .trailing)
            }
            .frame(maxWidth: .infinity)
            .padding(.leading, orderBookOuterPadding)
            .padding(.trailing, orderBookCenterPadding)

            HStack(spacing: 0) {
                SkeletonBlock(width: priceWidth(for: seed + 2), height: 13, cornerRadius: 3)
                    .frame(maxWidth: .infinity, alignment: .leading)
                SkeletonBlock(width: sizeWidth(for: seed + 1), height: 13, cornerRadius: 3)
                    .frame(maxWidth: .infinity, alignment: .trailing)
            }
            .frame(maxWidth: .infinity)
            .padding(.leading, orderBookCenterPadding)
            .padding(.trailing, orderBookOuterPadding)
        }
        .frame(height: 30)
    }

    private func priceWidth(for seed: Int) -> CGFloat {
        [72, 84, 64, 78][seed % 4]
    }

    private func sizeWidth(for seed: Int) -> CGFloat {
        [34, 46, 38, 52][seed % 4]
    }
}

private struct OrderBookList: View {
    let orderBook: OrderBookState

    var body: some View {
        LazyVStack(spacing: 0) {
            HeaderRow()

            SpreadRow(
                spreadText: orderBook.spreadText ?? "--",
                spreadPercentText: orderBook.spreadPercentText ?? "--"
            )

            ForEach(Array(orderBook.pairedRows.enumerated()), id: \.offset) { _, pair in
                PairedLevelRow(pair: pair)
            }
        }
    }
}

private struct HeaderRow: View {
    var body: some View {
        HStack(spacing: orderBookTextCenterGap) {
            HStack(spacing: 0) {
                HeaderCell(text: "Size", alignment: .leading)
                HeaderCell(text: "Price (Bid)", alignment: .trailing)
            }
            .frame(maxWidth: .infinity)
            .padding(.leading, orderBookOuterPadding)
            .padding(.trailing, orderBookCenterPadding)

            HStack(spacing: 0) {
                HeaderCell(text: "Price (Ask)", alignment: .leading)
                HeaderCell(text: "Size", alignment: .trailing)
            }
            .frame(maxWidth: .infinity)
            .padding(.leading, orderBookCenterPadding)
            .padding(.trailing, orderBookOuterPadding)
        }
        .padding(.vertical, 6)
    }
}

private struct HeaderCell: View {
    let text: String
    let alignment: Alignment

    var body: some View {
        Text(text)
            .font(.system(size: 11))
            .foregroundStyle(MR.colors.shared.text_tertiary.swiftUIColor)
            .frame(maxWidth: .infinity, alignment: alignment)
            .lineLimit(1)
            .minimumScaleFactor(0.75)
    }
}

private struct PairedLevelRow: View {
    let pair: OrderBookLevelPair

    var body: some View {
        ZStack {
            DepthColumns(pair: pair)

            HStack(spacing: orderBookTextCenterGap) {
                BidColumns(level: pair.bid)
                AskColumns(level: pair.ask)
            }
            .font(.system(size: 13, design: .monospaced))
            .lineLimit(1)
        }
        .frame(height: 30)
    }
}

private struct DepthColumns: View {
    let pair: OrderBookLevelPair

    var body: some View {
        HStack(spacing: orderBookDepthCenterGap) {
            HStack(spacing: 0) {
                PriceDepth(level: pair.bid, side: .bid)
            }
            .frame(maxWidth: .infinity)
            .padding(.leading, orderBookOuterPadding)
            .padding(.trailing, orderBookCenterPadding)

            HStack(spacing: 0) {
                PriceDepth(level: pair.ask, side: .ask)
            }
            .frame(maxWidth: .infinity)
            .padding(.leading, orderBookCenterPadding)
            .padding(.trailing, orderBookOuterPadding)
        }
    }
}

private struct PriceDepth: View {
    let level: OrderBookRow?
    let side: OrderBookSide

    var body: some View {
        GeometryReader { geometry in
            ZStack(alignment: side == .bid ? .trailing : .leading) {
                sideColor(for: side)
                    .opacity(level == nil ? 0 : 0.11)
                    .frame(width: geometry.size.width * CGFloat(depthFraction))
                    .animation(.easeInOut(duration: 0.18), value: depthFraction)
            }
            .frame(
                width: geometry.size.width,
                height: geometry.size.height,
                alignment: side == .bid ? .trailing : .leading
            )
        }
        .frame(maxWidth: .infinity)
        .frame(height: 30)
    }

    private func clamped(_ value: Double) -> Double {
        min(max(value, 0), 1)
    }

    private var depthFraction: Double {
        level.map { clamped($0.depthFraction) } ?? 0
    }
}

private struct BidColumns: View {
    let level: OrderBookRow?

    var body: some View {
        HStack(spacing: 0) {
            SizeText(level: level, side: .bid, alignment: .leading)
            LevelText(level?.priceText, color: MR.colors.shared.bid.swiftUIColor, alignment: .trailing)
        }
        .frame(maxWidth: .infinity)
        .padding(.leading, orderBookOuterPadding)
        .padding(.trailing, orderBookCenterPadding)
    }
}

private struct AskColumns: View {
    let level: OrderBookRow?

    var body: some View {
        HStack(spacing: 0) {
            LevelText(level?.priceText, color: MR.colors.shared.ask.swiftUIColor, alignment: .leading)
            SizeText(level: level, side: .ask, alignment: .trailing)
        }
        .frame(maxWidth: .infinity)
        .padding(.leading, orderBookCenterPadding)
        .padding(.trailing, orderBookOuterPadding)
    }
}

private struct SizeText: View {
    let level: OrderBookRow?
    let side: OrderBookSide
    let alignment: Alignment

    @State private var flashOpacity = 0.0

    var body: some View {
        ZStack(alignment: alignment) {
            Text(level?.sizeText ?? "")
                .foregroundStyle(MR.colors.shared.text_primary.swiftUIColor)

            Text(level?.sizeText ?? "")
                .foregroundStyle(sideColor(for: side).opacity(flashOpacity))
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: alignment)
        .lineLimit(1)
        .minimumScaleFactor(0.72)
        .onChange(of: flashTrigger) { _, _ in
            runFlash()
        }
    }

    private var flashTrigger: String? {
        guard let level,
              level.change != .none,
              level.sizeChangeFraction >= orderBookFlashMinChangeFraction
        else {
            return nil
        }
        return "\(level.rowKey):\(level.sizeText):\(level.change)"
    }

    private var flashStartOpacity: Double {
        switch level?.change {
        case .up:
            return orderBookFlashUpOpacity
        case .down:
            return orderBookFlashDownOpacity
        default:
            return 0
        }
    }

    private func runFlash() {
        let startOpacity = flashStartOpacity
        guard startOpacity > 0 else {
            flashOpacity = 0
            return
        }
        flashOpacity = startOpacity
        withAnimation(.easeOut(duration: 0.24)) {
            flashOpacity = 0
        }
    }
}

private struct LevelText: View {
    let text: String?
    let color: Color
    let alignment: Alignment

    init(_ text: String?, color: Color, alignment: Alignment) {
        self.text = text
        self.color = color
        self.alignment = alignment
    }

    var body: some View {
        Text(text ?? "")
            .foregroundStyle(color)
            .frame(maxWidth: .infinity, alignment: alignment)
            .lineLimit(1)
            .minimumScaleFactor(0.72)
    }
}

private struct SpreadRow: View {
    let spreadText: String
    let spreadPercentText: String

    var body: some View {
        HStack(spacing: orderBookSpreadItemGap) {
            Text("Spread")
                .font(.system(size: 12))
                .foregroundStyle(MR.colors.shared.text_tertiary.swiftUIColor)

            Text(spreadText)
                .font(.system(size: 15, weight: .semibold, design: .monospaced))
                .foregroundStyle(MR.colors.shared.brand_orange.swiftUIColor)

            Text(spreadPercentText)
                .font(.system(size: 12, weight: .medium, design: .monospaced))
                .foregroundStyle(MR.colors.shared.text_tertiary.swiftUIColor)
        }
        .lineLimit(1)
        .minimumScaleFactor(0.72)
        .frame(maxWidth: .infinity)
        .frame(height: 38)
        .padding(.horizontal, orderBookOuterPadding)
        .background(MR.colors.shared.app_panel.swiftUIColor)
    }
}

private let orderBookTextCenterGap: CGFloat = 8
private let orderBookDepthCenterGap: CGFloat = 0
private let orderBookCenterPadding: CGFloat = 0
private let orderBookOuterPadding: CGFloat = 4
private let orderBookSpreadItemGap: CGFloat = 12
private let orderBookFlashMinChangeFraction = 0.08
private let orderBookFlashUpOpacity = 0.88
private let orderBookFlashDownOpacity = 0.78

private func sideColor(for side: OrderBookSide) -> Color {
    side == .bid ? MR.colors.shared.bid.swiftUIColor : MR.colors.shared.ask.swiftUIColor
}
